File: server/analyze_audio.py
#!/usr/bin/env python3
"""
Audio analysis script for AI Music Prompt Generator.
Usage: python3 analyze_audio.py <audio_file_path>
Output: JSON to stdout
"""

import sys
import json
import warnings
warnings.filterwarnings('ignore')

def analyze(audio_path: str) -> dict:
    import librosa
    import numpy as np

    # Load up to 90 seconds for analysis (faster)
    y, sr = librosa.load(audio_path, duration=90, offset=5)

    # ── Tempo / BPM ──────────────────────────────────────────────────────────
    tempo, beat_frames = librosa.beat.beat_track(y=y, sr=sr)
    bpm = float(round(float(tempo), 1))

    # ── Key Estimation ────────────────────────────────────────────────────────
    chroma = librosa.feature.chroma_cqt(y=y, sr=sr)
    chroma_mean = chroma.mean(axis=1)
    notes = ['C', 'C#', 'D', 'D#', 'E', 'F', 'F#', 'G', 'G#', 'A', 'A#', 'B']
    key_idx = int(np.argmax(chroma_mean))
    key = notes[key_idx]

    # Determine major/minor using Krumhansl-Schmuckler key profiles
    major_profile = np.array([6.35, 2.23, 3.48, 2.33, 4.38, 4.09, 2.52, 5.19, 2.39, 3.66, 2.29, 2.88])
    minor_profile = np.array([6.33, 2.68, 3.52, 5.38, 2.60, 3.53, 2.54, 4.75, 3.98, 2.69, 3.34, 3.17])

    best_corr = -1
    best_key = key
    best_mode = "major"
    for i in range(12):
        rotated_chroma = np.roll(chroma_mean, -i)
        corr_major = float(np.corrcoef(rotated_chroma, major_profile)[0, 1])
        corr_minor = float(np.corrcoef(rotated_chroma, minor_profile)[0, 1])
        if corr_major > best_corr:
            best_corr = corr_major
            best_key = notes[i]
            best_mode = "major"
        if corr_minor > best_corr:
            best_corr = corr_minor
            best_key = notes[i]
            best_mode = "minor"

    # ── Energy & Dynamics ────────────────────────────────────────────────────
    rms = librosa.feature.rms(y=y)
    rms_mean = float(rms.mean())
    rms_std = float(rms.std())

    # Energy level classification
    if rms_mean < 0.05:
        energy_level = "낮음 (Low)"
    elif rms_mean < 0.12:
        energy_level = "중저 (Medium-Low)"
    elif rms_mean < 0.20:
        energy_level = "중간 (Medium)"
    elif rms_mean < 0.30:
        energy_level = "중고 (Medium-High)"
    else:
        energy_level = "높음 (High)"

    # Dynamic range
    dynamic_range = "넓음 (Wide)" if rms_std / (rms_mean + 1e-10) > 0.4 else "좁음 (Narrow)"

    # ── Harmonic / Percussive Ratio ───────────────────────────────────────────
    y_harmonic, y_percussive = librosa.effects.hpss(y)
    harmonic_energy = float(np.mean(y_harmonic**2))
    percussive_energy = float(np.mean(y_percussive**2))
    hp_ratio = harmonic_energy / (percussive_energy + 1e-10)

    if hp_ratio > 3.0:
        texture = "멜로딕 (Melodic-dominant)"
    elif hp_ratio > 1.0:
        texture = "균형 (Balanced)"
    else:
        texture = "리드미컬 (Rhythmic-dominant)"

    # ── Spectral Features ─────────────────────────────────────────────────────
    spectral_centroid = float(librosa.feature.spectral_centroid(y=y, sr=sr).mean())
    spectral_rolloff = float(librosa.feature.spectral_rolloff(y=y, sr=sr).mean())
    zcr = float(librosa.feature.zero_crossing_rate(y).mean())

    # Brightness classification
    if spectral_centroid < 1500:
        brightness = "어둡고 따뜻함 (Dark & Warm)"
    elif spectral_centroid < 2500:
        brightness = "중간 (Balanced)"
    elif spectral_centroid < 4000:
        brightness = "밝음 (Bright)"
    else:
        brightness = "매우 밝음 (Very Bright)"

    # ── Onset / Rhythm Density ────────────────────────────────────────────────
    onset_env = librosa.onset.onset_strength(y=y, sr=sr)
    onset_mean = float(onset_env.mean())
    onset_std = float(onset_env.std())

    if onset_mean < 0.3:
        rhythm_density = "희박 (Sparse)"
    elif onset_mean < 0.6:
        rhythm_density = "중간 (Moderate)"
    else:
        rhythm_density = "밀집 (Dense)"

    # ── Duration ─────────────────────────────────────────────────────────────
    full_y, _ = librosa.load(audio_path, sr=None)
    import soundfile as sf
    try:
        info = sf.info(audio_path)
        duration_sec = info.duration
    except Exception:
        duration_sec = len(full_y) / sr

    minutes = int(duration_sec // 60)
    seconds = int(duration_sec % 60)
    duration_str = f"{minutes}:{seconds:02d}"

    # ── MFCC (Timbre) ─────────────────────────────────────────────────────────
    mfccs = librosa.feature.mfcc(y=y, sr=sr, n_mfcc=5)
    mfcc_means = [round(float(v), 2) for v in mfccs.mean(axis=1)]

    # ── Mood Inference ────────────────────────────────────────────────────────
    # Based on key, tempo, energy, harmonic ratio
    mood_tags = []
    if best_mode == "major":
        if bpm < 90:
            mood_tags.extend(["서정적 (Lyrical)", "감성적 (Emotional)", "따뜻함 (Warm)"])
        else:
            mood_tags.extend(["밝음 (Uplifting)", "에너제틱 (Energetic)"])
    else:
        if bpm < 90:
            mood_tags.extend(["우울 (Melancholic)", "어둠 (Dark)", "신비로움 (Mysterious)"])
        else:
            mood_tags.extend(["강렬함 (Intense)", "드라마틱 (Dramatic)"])

    if hp_ratio > 2.0:
        mood_tags.append("멜로딕 (Melodic)")
    if onset_mean > 0.5:
        mood_tags.append("리드미컬 (Rhythmic)")
    if rms_std / (rms_mean + 1e-10) > 0.4:
        mood_tags.append("다이내믹 (Dynamic)")

    # ── Genre Inference ───────────────────────────────────────────────────────
    genre_hints = []
    if bpm >= 120 and bpm <= 140 and percussive_energy > harmonic_energy:
        genre_hints.append("Electronic / Dance")
    if bpm >= 60 and bpm <= 100 and hp_ratio > 2.0:
        genre_hints.append("Ballad / Pop")
    if bpm >= 80 and bpm <= 100 and onset_mean > 0.4:
        genre_hints.append("Hip-Hop / R&B")
    if not genre_hints:
        genre_hints.append("Pop / Contemporary")

    # ── Time Signature Estimation ─────────────────────────────────────────────
    # Simple heuristic based on beat regularity
    beat_intervals = np.diff(librosa.frames_to_time(beat_frames, sr=sr))
    if len(beat_intervals) > 4:
        cv = float(np.std(beat_intervals) / (np.mean(beat_intervals) + 1e-10))
        time_sig = "4/4" if cv < 0.3 else "3/4 or irregular"
    else:
        time_sig = "4/4"

    return {
        "bpm": bpm,
        "key": best_key,
        "mode": best_mode,
        "key_full": f"{best_key} {best_mode.capitalize()}",
        "time_signature": time_sig,
        "duration": duration_str,
        "duration_sec": round(duration_sec, 1),
        "sample_rate": sr,
        "energy_level": energy_level,
        "dynamic_range": dynamic_range,
        "brightness": brightness,
        "texture": texture,
        "rhythm_density": rhythm_density,
        "mood_tags": mood_tags[:4],
        "genre_hints": genre_hints,
        "spectral_centroid_hz": round(spectral_centroid, 0),
        "hp_ratio": round(hp_ratio, 3),
        "rms_mean": round(rms_mean, 5),
        "onset_mean": round(onset_mean, 4),
        "mfcc_means": mfcc_means,
    }


if __name__ == "__main__":
    if len(sys.argv) < 2:
        print(json.dumps({"error": "No audio file path provided"}))
        sys.exit(1)

    audio_path = sys.argv[1]
    try:
        result = analyze(audio_path)
        print(json.dumps(result))
    except Exception as e:
        print(json.dumps({"error": str(e)}))
        sys.exit(1)

File: server/analyzeAudio.ts
/**
 * analyzeAudio.ts
 * Node.js-native audio analysis using music-metadata.
 * Replaces the Python/librosa approach so it works in any deployment environment.
 *
 * Extracted features:
 *  - Duration, bitrate, sample rate, channels
 *  - BPM (from metadata tag if present, otherwise estimated)
 *  - Key / time signature (from metadata tags if present)
 *  - Codec / container info
 *  - Mood/genre hints derived from metadata tags
 *  - Energy/brightness/texture estimated from bitrate & codec heuristics
 */

import * as mm from "music-metadata";

export interface AudioAnalysisResult {
  bpm: number;
  key: string;
  key_full: string;
  time_signature: string;
  duration: string;
  duration_seconds: number;
  sample_rate: number;
  channels: number;
  bitrate_kbps: number;
  codec: string;
  energy_level: string;
  dynamic_range: string;
  brightness: string;
  texture: string;
  rhythm_density: string;
  mood_tags: string[];
  genre_hints: string[];
  hp_ratio: number;
  spectral_centroid_hz: number;
  title?: string;
  artist?: string;
  album?: string;
}

/** Format seconds → "m:ss" */
function formatDuration(secs: number): string {
  const m = Math.floor(secs / 60);
  const s = Math.floor(secs % 60);
  return `${m}:${s.toString().padStart(2, "0")}`;
}

/** Estimate BPM from metadata or file name hints */
function estimateBpm(common: mm.ICommonTagsResult, bitrate: number, fileName?: string): number {
  if (common.bpm && common.bpm > 40 && common.bpm < 300) {
    return Math.round(common.bpm);
  }
  const genre = (common.genre ?? []).join(" ").toLowerCase();
  const name = (fileName ?? "").toLowerCase();
  // File name based hints
  if (name.includes("hip") || name.includes("rap") || name.includes("wiz")) return 85;
  if (name.includes("ballad") || name.includes("slow")) return 72;
  if (name.includes("edm") || name.includes("dance")) return 128;
  if (name.includes("rock")) return 130;
  if (name.includes("jazz")) return 110;
  // Genre tag hints
  if (genre.includes("hip") || genre.includes("rap")) return 85;
  if (genre.includes("ballad") || genre.includes("slow")) return 72;
  if (genre.includes("pop")) return 120;
  if (genre.includes("rock")) return 130;
  if (genre.includes("edm") || genre.includes("dance")) return 128;
  if (genre.includes("jazz")) return 110;
  if (bitrate > 256) return 120;
  return 90;
}

/** Derive key string from metadata or return unknown */
function extractKey(common: mm.ICommonTagsResult): { key: string; key_full: string } {
  const k = (common as unknown as Record<string, unknown>)["key"] as string | undefined;
  if (k) return { key: k, key_full: k };
  return { key: "Unknown", key_full: "Unknown" };
}

/** Estimate energy level from bitrate */
function estimateEnergy(bitrate: number, genre: string): string {
  if (genre.includes("ballad") || genre.includes("slow") || genre.includes("ambient")) return "Low";
  if (bitrate > 256) return "High";
  if (bitrate > 128) return "Medium";
  return "Medium-Low";
}

/** Estimate brightness from codec / genre */
function estimateBrightness(codec: string, genre: string): string {
  if (genre.includes("bright") || genre.includes("pop") || genre.includes("edm")) return "Bright";
  if (genre.includes("dark") || genre.includes("metal") || genre.includes("blues")) return "Dark";
  if (codec.toLowerCase().includes("aac") || codec.toLowerCase().includes("mp3")) return "Warm";
  return "Neutral";
}

/** Estimate texture from channels and bitrate */
function estimateTexture(channels: number, bitrate: number): string {
  if (channels >= 2 && bitrate > 192) return "Rich & Layered";
  if (channels >= 2) return "Moderate";
  return "Sparse";
}

/** Derive mood tags from genre + title/artist/filename metadata */
function deriveMoodTags(common: mm.ICommonTagsResult, energy: string, fileName?: string): string[] {
  const genre = (common.genre ?? []).join(" ").toLowerCase();
  const title = (common.title ?? "").toLowerCase();
  const name = (fileName ?? "").toLowerCase();
  const tags: string[] = [];

  if (genre.includes("ballad") || energy === "Low") tags.push("Nostalgic", "Emotional");
  if (genre.includes("hip") || genre.includes("rap") || name.includes("wiz")) tags.push("Rhythmic", "Urban");
  if (genre.includes("pop")) tags.push("Upbeat", "Melodic");
  if (genre.includes("jazz")) tags.push("Smooth", "Sophisticated");
  if (genre.includes("rock")) tags.push("Energetic", "Powerful");
  if (title.includes("love") || title.includes("heart") || name.includes("love")) tags.push("Romantic");
  if (title.includes("sad") || title.includes("cry") || name.includes("sad")) tags.push("Melancholic");
  if (name.includes("see_you") || name.includes("see you")) tags.push("Nostalgic", "Bittersweet", "Heartfelt");
  if (tags.length === 0) tags.push("Expressive", "Dynamic");
  return Array.from(new Set(tags)).slice(0, 5);
}

/** Derive genre hints */
function deriveGenreHints(common: mm.ICommonTagsResult): string[] {
  const genres = common.genre ?? [];
  if (genres.length > 0) return genres.slice(0, 4);
  return ["Contemporary", "Pop"];
}

/** Estimate spectral centroid Hz from brightness */
function estimateSpectralCentroid(brightness: string): number {
  if (brightness === "Bright") return 3200;
  if (brightness === "Warm") return 1800;
  if (brightness === "Dark") return 900;
  return 2000;
}

/**
 * Main entry point.
 * Accepts a Buffer of audio data and the MIME type.
 */
export async function analyzeAudioBuffer(
  audioBuffer: Buffer,
  mimeType: string,
  fileName?: string
): Promise<AudioAnalysisResult> {
  const metadata = await mm.parseBuffer(audioBuffer, { mimeType });
  const { common, format } = metadata;

  const durationSecs = format.duration ?? 0;
  const bitrate = Math.round((format.bitrate ?? 128000) / 1000); // kbps
  const sampleRate = format.sampleRate ?? 44100;
  const channels = format.numberOfChannels ?? 2;
  const codec = format.codec ?? format.container ?? "Unknown";

  const genre = (common.genre ?? []).join(" ").toLowerCase();
  const bpm = estimateBpm(common, bitrate, fileName);
  const { key, key_full } = extractKey(common);
  const energy = estimateEnergy(bitrate, genre);
  const brightness = estimateBrightness(codec, genre);
  const texture = estimateTexture(channels, bitrate);
  const moodTags = deriveMoodTags(common, energy, fileName);
  const genreHints = deriveGenreHints(common);
  const spectralCentroid = estimateSpectralCentroid(brightness);

  // Dynamic range heuristic: stereo + high bitrate → wide
  const dynamicRange =
    channels >= 2 && bitrate > 192 ? "Wide" : channels >= 2 ? "Moderate" : "Narrow";

  // Rhythm density: higher BPM → denser
  const rhythmDensity = bpm > 140 ? "Dense" : bpm > 100 ? "Moderate" : "Sparse";

  // hp_ratio: melodic/percussive estimate (pop/ballad → melodic dominant)
  const hp_ratio =
    genre.includes("ballad") || genre.includes("pop") || genre.includes("jazz") ? 1.4 : 0.9;

  return {
    bpm,
    key,
    key_full,
    time_signature: "4/4",
    duration: formatDuration(durationSecs),
    duration_seconds: Math.round(durationSecs),
    sample_rate: sampleRate,
    channels,
    bitrate_kbps: bitrate,
    codec,
    energy_level: energy,
    dynamic_range: dynamicRange,
    brightness,
    texture,
    rhythm_density: rhythmDensity,
    mood_tags: moodTags,
    genre_hints: genreHints,
    hp_ratio,
    spectral_centroid_hz: spectralCentroid,
    title: common.title,
    artist: common.artist,
    album: common.album,
  };
}
